transportable-udfs-spark/src/main/scala/com/linkedin/transport/spark/data/SparkMap.scala

```diff
@@ -31,10 +31,14 @@ case class SparkMap(private var _mapData: MapData,
   }
 
   override def keySet(): util.Set[StdData] = {
+    if (_mutableMap == null) {
+      _mutableMap = createMutableMap()
+    }
```

Comment:
Creating a mutable map on the read path seems excessive. I see that `_map.keyArray()` exposes getter methods. Can we build an iterator on top of this method ourselves? I think we would need to just maintain a current index inside the iterator and then increment it on `next()`.

Reply:
It is definitely excessive in this scenario, but I don't think there will be keySet/values calls without some get(key) call which would create a mutable map anyways.

I can make the change to make another Iterator though.

Reply:
I made the change to the iterator. It's a cleaner solution, thanks for suggestion!

Reply:
Nice! Looks much better. Consider reducing duplicated code? Something like this?
```scala
  override def keySet(): util.Set[StdData] = {
    val keysIterator: Iterator[Any] = if (_mutableMap == null) {
      new Iterator[Any] {
        private val keyArray = _mapData.keyArray()
        var offset : Int = 0

        override def next(): Any = {
          offset += 1
          keyArray.get(offset -1, _keyType)
        }

        override def hasNext: Boolean = offset < SparkMap.this.size()
      }
    } else {
      _mutableMap.keysIterator
    }

    new util.AbstractSet[StdData] {
      override def iterator(): util.Iterator[StdData] = new util.Iterator[StdData] {

        override def next(): StdData = SparkWrapper.createStdData(keysIterator.next(), _keyType)

        override def hasNext: Boolean = keysIterator.hasNext
      }

      override def size(): Int = SparkMap.this.size()
    }
  }
```

Reply:
acked and updated.